File: packages/SystemUI/src/com/android/systemui/bubbles/BubbleOverflow.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.bubbles;

import static android.view.Display.INVALID_DISPLAY;
import static android.view.View.GONE;

import static com.android.systemui.bubbles.BadgedImageView.DEFAULT_PATH_SIZE;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Path;
import android.graphics.drawable.AdaptiveIconDrawable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.InsetDrawable;
import android.util.PathParser;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.android.systemui.R;

/**
 * Class for showing aged out bubbles.
 */
public class BubbleOverflow implements BubbleViewProvider {
    public static final String KEY = "Overflow";

    private BadgedImageView mOverflowBtn;
    private BubbleExpandedView mExpandedView;
    private LayoutInflater mInflater;
    private Context mContext;
    private Bitmap mIcon;
    private Path mPath;
    private int mBitmapSize;
    private int mIconBitmapSize;
    private int mDotColor;

    public BubbleOverflow(Context context) {
        mContext = context;
        mInflater = LayoutInflater.from(context);
        mBitmapSize = mContext.getResources().getDimensionPixelSize(R.dimen.bubble_bitmap_size);
        mIconBitmapSize = mContext.getResources().getDimensionPixelSize(
                R.dimen.bubble_overflow_icon_bitmap_size);
    }

    void setUpOverflow(ViewGroup parentViewGroup, BubbleStackView stackView) {
        mExpandedView = (BubbleExpandedView) mInflater.inflate(
                R.layout.bubble_expanded_view, parentViewGroup /* root */,
                false /* attachToRoot */);
        mExpandedView.setOverflow(true);
        mExpandedView.setStackView(stackView);

        updateIcon(mContext, parentViewGroup);
    }

    // TODO(b/149146374) Propagate theme change to bubbles in overflow.
    void updateIcon(Context context, ViewGroup parentViewGroup) {
        mInflater = LayoutInflater.from(context);
        mOverflowBtn = (BadgedImageView) mInflater.inflate(R.layout.bubble_overflow_button,
                parentViewGroup /* root */,
                false /* attachToRoot */);

        TypedArray ta = mContext.obtainStyledAttributes(
                new int[]{android.R.attr.colorBackgroundFloating});
        int bgColor = ta.getColor(0, Color.WHITE /* default */);
        ta.recycle();

        TypedValue typedValue = new TypedValue();
        context.getTheme().resolveAttribute(android.R.attr.colorAccent, typedValue, true);
        int colorAccent = mContext.getColor(typedValue.resourceId);
        mOverflowBtn.getDrawable().setTint(colorAccent);
        mDotColor = colorAccent;

        ColorDrawable bg = new ColorDrawable(bgColor);
        InsetDrawable fg = new InsetDrawable(mOverflowBtn.getDrawable(),
                mBitmapSize - mIconBitmapSize /* inset */);
        AdaptiveIconDrawable adaptiveIconDrawable = new AdaptiveIconDrawable(bg, fg);

        BubbleIconFactory iconFactory = new BubbleIconFactory(context);
        mIcon = iconFactory.createBadgedIconBitmap(adaptiveIconDrawable,
                null /* user */,
                true /* shrinkNonAdaptiveIcons */).icon;

        float scale = iconFactory.getNormalizer().getScale(mOverflowBtn.getDrawable(),
                null /* outBounds */, null /* path */, null /* outMaskShape */);
        float radius = DEFAULT_PATH_SIZE / 2f;
        mPath = PathParser.createPathFromPathData(
                context.getResources().getString(com.android.internal.R.string.config_icon_mask));
        Matrix matrix = new Matrix();
        matrix.setScale(scale /* x scale */, scale /* y scale */, radius /* pivot x */,
                radius /* pivot y */);
        mPath.transform(matrix);

        mOverflowBtn.setVisibility(GONE);
        mOverflowBtn.setRenderedBubble(this);
    }

    ImageView getBtn() {
        return mOverflowBtn;
    }

    void setBtnVisible(int visible) {
        mOverflowBtn.setVisibility(visible);
    }

    @Override
    public BubbleExpandedView getExpandedView() {
        return mExpandedView;
    }

    @Override
    public int getDotColor() {
        return mDotColor;
    }

    @Override
    public Bitmap getBadgedImage() {
        return mIcon;
    }

    @Override
    public boolean showDot() {
        return false;
    }

    @Override
    public Path getDotPath() {
        return mPath;
    }

    @Override
    public void setContentVisibility(boolean visible) {
        mExpandedView.setContentVisibility(visible);
    }

    @Override
    public void logUIEvent(int bubbleCount, int action, float normalX, float normalY,
            int index) {
        // TODO(b/149133814) Log overflow UI events.
    }

    @Override
    public View getIconView() {
        return mOverflowBtn;
    }

    @Override
    public String getKey() {
        return BubbleOverflow.KEY;
    }

    @Override
    public int getDisplayId() {
        return mExpandedView != null ? mExpandedView.getVirtualDisplayId() : INVALID_DISPLAY;
    }
}

File: packages/SystemUI/src/com/android/systemui/bubbles/BubbleOverflow.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.bubbles

import android.content.Context
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.Matrix
import android.graphics.Path
import android.graphics.drawable.AdaptiveIconDrawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.InsetDrawable
import android.util.PathParser
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import com.android.systemui.R

class BubbleOverflow(
    private val context: Context,
    private val stack: BubbleStackView
) : BubbleViewProvider {

    private var bitmap: Bitmap? = null
    private var dotPath: Path? = null
    private var bitmapSize = 0
    private var iconBitmapSize = 0
    private var dotColor = 0

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val expandedView: BubbleExpandedView = inflater
        .inflate(R.layout.bubble_expanded_view, null /* root */, false /* attachToRoot */)
            as BubbleExpandedView
    private val overflowBtn: BadgedImageView = inflater
        .inflate(R.layout.bubble_overflow_button, null /* root */, false /* attachToRoot */)
            as BadgedImageView
    init {
        updateResources()
        with(expandedView) {
            setOverflow(true)
            setStackView(stack)
            applyThemeAttrs()
        }
        with(overflowBtn) {
            setContentDescription(context.resources.getString(
                R.string.bubble_overflow_button_content_description))
            updateBtnTheme()
        }
    }

    fun update() {
        updateResources()
        expandedView.applyThemeAttrs()
        // Apply inset and new style to fresh icon drawable.
        overflowBtn.setImageResource(R.drawable.ic_bubble_overflow_button)
        updateBtnTheme()
    }

    fun updateResources() {
        bitmapSize = context.resources.getDimensionPixelSize(R.dimen.bubble_bitmap_size)
        iconBitmapSize = context.resources.getDimensionPixelSize(
                R.dimen.bubble_overflow_icon_bitmap_size)
        val bubbleSize = context.resources.getDimensionPixelSize(R.dimen.individual_bubble_size)
        overflowBtn.setLayoutParams(FrameLayout.LayoutParams(bubbleSize, bubbleSize))
        expandedView.updateDimensions()
    }

    fun updateBtnTheme() {
        val res = context.resources

        // Set overflow button accent color, dot color
        val typedValue = TypedValue()
        context.theme.resolveAttribute(android.R.attr.colorAccent, typedValue, true)

        val colorAccent = res.getColor(typedValue.resourceId)
        overflowBtn.getDrawable()?.setTint(colorAccent)
        dotColor = colorAccent

        // Set button and activity background color
        val nightMode = (res.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
                == Configuration.UI_MODE_NIGHT_YES)
        val bg = ColorDrawable(res.getColor(
                if (nightMode) R.color.bubbles_dark else R.color.bubbles_light))

        // Set button icon
        val iconFactory = BubbleIconFactory(context)
        val fg = InsetDrawable(overflowBtn.getDrawable(),
                bitmapSize - iconBitmapSize /* inset */)
        bitmap = iconFactory.createBadgedIconBitmap(AdaptiveIconDrawable(bg, fg),
                null /* user */, true /* shrinkNonAdaptiveIcons */).icon

        // Set dot path
        dotPath = PathParser.createPathFromPathData(
                res.getString(com.android.internal.R.string.config_icon_mask))
        val scale = iconFactory.normalizer.getScale(overflowBtn.getDrawable(),
                null /* outBounds */, null /* path */, null /* outMaskShape */)
        val radius = BadgedImageView.DEFAULT_PATH_SIZE / 2f
        val matrix = Matrix()
        matrix.setScale(scale /* x scale */, scale /* y scale */, radius /* pivot x */,
                radius /* pivot y */)
        dotPath?.transform(matrix)
        overflowBtn.setRenderedBubble(this)
    }

    fun setVisible(visible: Int) {
        overflowBtn.visibility = visible
    }

    override fun getExpandedView(): BubbleExpandedView? {
        return expandedView
    }

    override fun getDotColor(): Int {
        return dotColor
    }

    override fun getBadgedImage(): Bitmap? {
        return bitmap
    }

    override fun showDot(): Boolean {
        return false
    }

    override fun getDotPath(): Path? {
        return dotPath
    }

    override fun setContentVisibility(visible: Boolean) {
        expandedView.setContentVisibility(visible)
    }

    override fun getIconView(): View? {
        return overflowBtn
    }

    override fun getKey(): String {
        return KEY
    }

    override fun getDisplayId(): Int {
        return expandedView.virtualDisplayId
    }

    companion object {
        @JvmField val KEY = "Overflow"
    }
}